compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -13314,113 +13577,92 @@ OMR::Z::TreeEvaluator::primitiveArraycopyEvaluator(TR::Node* node, TR::CodeGener
-      TR_ASSERT_FATAL(TR::Compiler->target.is64Bit(),
-                 "GSCS: Guarded Load OOL Path only defined in 64 bit mode");
-      TR::Register *vmReg = cg->getMethodMetaDataRealRegister();
-      TR::MemoryReference *privFlagMR = generateS390MemoryReference(vmReg,
-            TR::Compiler->vm.thisThreadGetConcurrentScavengeActiveByteAddressOffset(cg->comp()), cg);
-      generateSIInstruction(cg, TR::InstOpCode::TM, node, privFlagMR, 0x00000002);
-      generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_CC3, node, slowPathLabel);
-
-      // Generate OOL Slow Path
-      TR_S390OutOfLineCodeSection* outOfLineCodeSection = new (cg->trHeapMemory()) TR_S390OutOfLineCodeSection(slowPathLabel, mergeLabel, cg);
-      cg->getS390OutOfLineCodeSectionList().push_front(outOfLineCodeSection);
-      outOfLineCodeSection->swapInstructionListsWithCompilation();
-
-      generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, slowPathLabel);
-
-      TR::Register* strideRegister = cg->allocateRegister();
-
-      MemCpyVarLenTypedMacroOp vtOpSlow(node, byteDstNode, byteSrcNode, cg, node->getArrayCopyElementType(), byteLenReg, byteLenNode, true, node->isForwardArrayCopy());
-      vtOpSlow.generate(byteDstReg, byteSrcReg, strideRegister);
-      vtOpSlow.cleanUpReg();
-
-      cg->stopUsingRegister(strideRegister);
-
-      generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, mergeLabel);
-
-      // Switch back from OOL Slow Path
-      outOfLineCodeSection->swapInstructionListsWithCompilation();
+      generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, startICFLabel);
+      startICFLabel->setStartInternalControlFlow();
+      cursor = generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::getCmpOpCode(), node, byteLenReg, static_cast<int8_t>(0x0), TR::InstOpCode::COND_BNH, mergeLabel, false);
+      iComment("byteLen <= 0 goto mergeLabel");
       }
-#endif
 
    if (node->isForwardArrayCopy())
       {
-      if (byteLenNode->getOpCode().isLoadConst())
-         {
-         MemCpyConstLenMacroOp op(node, byteDstNode, byteSrcNode, cg, byteLenNode->getConst<int64_t>());
-
-         op.generate(byteDstReg, byteSrcReg);
-         }
-      else
-         {
-         bool alreadyEvaluatedLengthMinusOne = false;
-
-         // Don't evaluate length minus one if byteLenNode was already evaluated
-         if (!alreadyEvaluatedByteLenNode)
-            {
-            byteLenReg = cg->evaluateLengthMinusOneForMemoryOps(byteLenNode, true, alreadyEvaluatedLengthMinusOne);
-            }
-
-         MemCpyVarLenMacroOp op(node, byteDstNode, byteSrcNode, cg, byteLenReg, byteLenNode, alreadyEvaluatedLengthMinusOne);
-
-         op.setUseEXForRemainder(true);
-         op.generate(byteDstReg, byteSrcReg);
-         }
+      // Optimizer has already identified that this is array copy in forward direction.
```

Comment:
I don't think this comment is true or useful for that matter. It could have very well been the IL generator from whichever downstream project is consuming OMR which sets this flag. I think we should remove the comment as it serves no purpose explaining _why_ the code is the way it is. i.e. it doesn't give any more information than the `if (node->isForwardArrayCopy())` line proceeding it.